tensorflow/lite/micro/tools/requantize_flatbuffer_utils.py

```diff
@@ -0,0 +1,215 @@
+    2: np.int32,
+    3: np.uint8,
+    4: np.int64,
+    5: np.string_,
+    6: np.bool_,
+    7: np.int16,
+    8: np.complex64,
+    9: np.int8,
+    10: np.float64,
+    11: np.complex128,
+    12: np.uint64,
+    13: "RESORCE",
+    14: "VARIANT",
+    15: "UINT32",
+    16: "UINT16",
+    17: "INT4",
+}
+
+TENSOR_TYPE_CODE = dict((reversed(item) for item in TENSOR_CODE_TYPE.items()))
+"""
+Quantization utils functions
+"""
+
+
+def clip_range(vals, bit_width):
+  """Mimic integer calculation.
+
+  Clip the range of vals based on bit width.
+
+  e.g., clip_range([300], 8) = [127] since int8 have range [-128, 127]
+
+  Args:
+      vals (np.array): float representation of the integer values
+      bit_width (int): number of desired bits for vals
+
+  Returns:
+      np.array : clipped vals
+  """
+  # Numpy integer calculation does not do saturation. Implement here
+  half = np.power(2.0, bit_width) / 2
+  min_val, max_val = -half, half - 1
+  if vals.max() > max_val or vals.min() < min_val:
+    logging.info(f"WARNING: integer overflow!")
+  return np.round(np.clip(vals, min_val, max_val))
+
+
+def quantize_data(data, scale, zero_point=0, bit_width=8):
+  """Quantize the data to integer type with desired bit width.
+
+  The quantized data is represented using float since integer calculation in
+  numpy may differ from other implementations (e.g., no integer saturation
+  protection in numpy)
+
+  Args:
+      data (np.array): float data
+      scale (float): quantization scale of the data
+      zero_point (integer): quantization zero point of the data
+      bit_width (int): number of representative bits for vals
+
+  Returns:
+      np.array : quantized data in float but clipped range
+  """
+  vals = np.round(data / scale) + zero_point
+  return clip_range(vals, bit_width)
+
+
+def dequantize_data(quantized_data, scale, zero_point=0):
+  """Dequantize the data to integer type with desired bit width.
+
+  Args:
+      quantized_data (np.array): quantized data
+      scale (float): quantization scale of the data
+      zero_point (integer): quantization zero point of the data
+
+  Returns:
+      np.array : dequantized data
+  """
+  return scale * (quantized_data - zero_point)
+
+
+'''Conversion Util Functions'''
+
+
+def change_quantization_settings_8to16(tensor):
+  if not tensor.quantization:
+    return
+  assert (tensor.quantization.quantizedDimension == 0
+          )  # Only layer quantization supported
+  scale, zero_point = (
+      tensor.quantization.scale[0],
+      tensor.quantization.zeroPoint[0],
+  )
+  # Asymmertical quantized: scale * (qmax - zero_point) = rmax
+  rmax = scale * (MAX_INT8 - zero_point)
+  rmin = scale * (MIN_INT8 - zero_point)
+  # symmertical quantized: scale * qmax = rmax
+  scale_16 = max(abs(rmax), abs(rmin)) / abs(MIN_INT16)
+  # Change scale: Symmetrical Quantized
+  tensor.quantization.scale = [scale_16]
+  tensor.quantization.zeroPoint = [0]
+
+
+def change_activation_tensor_8to16(tensor):
```

Comment:
add comment about what the function is doing?

Reply:
Done.